File: crm_dashboard/src/lib/activity-text.test.ts
import { describe, expect, it } from "vitest";
import { activityToTimelineEntry } from "./activity-text";
import type { Activity } from "./activities";

const NAMES = new Map([
  ["m-budi", "Budi Santoso"],
  ["m-sari", "Sari Wulandari"],
]);

function activity(overrides: Partial<Activity>): Activity {
  return {
    id: "a-1",
    lead_id: "l-1",
    type: "note_added",
    actor_membership_id: null,
    body: null,
    metadata: null,
    created_at: "2026-08-17T09:00:00Z",
    ...overrides,
  };
}

describe("activityToTimelineEntry", () => {
  // lead_created: internal/lead/usecase.go's Create passes metadata=nil.
  it("lead_created — no metadata, static text", () => {
    const entry = activityToTimelineEntry(activity({ type: "lead_created", metadata: null }), NAMES);
    expect(entry).toEqual({ isHuman: false, text: "Lead dibuat" });
  });

  // status_changed: {"from": "...", "to": "..."} — internal/lead/usecase.go UpdateStatus.
  it("status_changed — renders from -> to using status labels, not raw enum values", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "status_changed", metadata: { from: "new", to: "contacted" } }),
      NAMES
    );
    expect(entry.text).toBe("Status: Baru → Dihubungi");
  });

  // lead_assigned: {"from": <id|null>, "to": <id>} — internal/lead/usecase.go UpdateAssignment.
  it("lead_assigned — first-time assignment (no `from`) reads as \"ditugaskan ke\"", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "lead_assigned", metadata: { to: "m-budi" } }),
      NAMES
    );
    expect(entry.text).toBe("Ditugaskan ke Budi Santoso");
  });

  it("lead_assigned — reassignment (both from and to present) reads as a transfer", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "lead_assigned", metadata: { from: "m-budi", to: "m-sari" } }),
      NAMES
    );
    expect(entry.text).toBe("Dipindahkan dari Budi Santoso ke Sari Wulandari");
  });

  // lead_unassigned: {"from": <id>} — internal/lead/usecase.go UpdateAssignment.
  it("lead_unassigned", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "lead_unassigned", metadata: { from: "m-budi" } }),
      NAMES
    );
    expect(entry.text).toBe("Dilepas dari Budi Santoso");
  });

  it("resolves a membership id no longer in the map as a deactivated member, not a crash", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "lead_assigned", metadata: { to: "m-gone" } }),
      NAMES
    );
    expect(entry.text).toBe("Ditugaskan ke Anggota yang sudah tidak aktif");
  });

  // lead_converted: {"customer_id": "..."} — internal/customer/usecase.go Convert.
  it("lead_converted", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "lead_converted", metadata: { customer_id: "c-1" } }),
      NAMES
    );
    expect(entry.text).toBe("Dikonversi menjadi customer");
  });

  // task_created: {"task_id": "...", "title": "..."} — internal/task/usecase.go Create.
  it("task_created — includes the title", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "task_created", metadata: { task_id: "t-1", title: "Telepon lagi besok" } }),
      NAMES
    );
    expect(entry.text).toBe("Task dibuat: Telepon lagi besok");
  });

  // task_completed: {"task_id": "..."} ONLY — no title in this metadata,
  // unlike task_created. A naive implementation might assume it's there.
  it("task_completed — has no title in its metadata, must not crash or show \"undefined\"", () => {
    const entry = activityToTimelineEntry(
      activity({ type: "task_completed", metadata: { task_id: "t-1" } }),
      NAMES
    );
    expect(entry.text).toBe("Task diselesaikan");
  });

  it.each(["note_added", "call_logged", "whatsapp_opened"] as const)(
    "%s — human-authored, body is the text, author resolved from actor_membership_id",
    (type) => {
      const entry = activityToTimelineEntry(
        activity({ type, body: "Sudah dihubungi, tertarik.", actor_membership_id: "m-sari" }),
        NAMES
      );
      expect(entry.isHuman).toBe(true);
      expect(entry.text).toBe("Sudah dihubungi, tertarik.");
      expect(entry.authorName).toBe("Sari Wulandari");
    }
  );
});

File: crm_employee/test/features/leads/presentation/activity_text_test.dart
import 'package:crm_employee/features/leads/domain/entities/activity.dart';
import 'package:crm_employee/features/leads/presentation/activity_text.dart';
import 'package:flutter_test/flutter_test.dart';

Activity _activity({
  String type = 'note_added',
  String? actorMembershipId,
  String? body,
  Map<String, dynamic>? metadata,
}) {
  return Activity(
    id: 'activity-1',
    leadId: 'lead-1',
    type: type,
    actorMembershipId: actorMembershipId,
    body: body,
    metadata: metadata,
    createdAt: DateTime.utc(2026, 8, 30),
  );
}

const _me = 'membership-me';
const _other = 'membership-other';

void main() {
  group('activityToTimelineEntry', () {
    test('lead_created renders a static line, ignores metadata', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'lead_created'),
        myMembershipId: _me,
      );
      expect(entry.isHuman, isFalse);
      expect(entry.text, 'Lead dibuat');
    });

    test('status_changed renders both labels from metadata', () {
      final entry = activityToTimelineEntry(
        _activity(
          type: 'status_changed',
          metadata: {'from': 'new', 'to': 'contacted'},
        ),
        myMembershipId: _me,
      );
      expect(entry.isHuman, isFalse);
      expect(entry.text, 'Status: Baru → Dihubungi');
    });

    test('status_changed falls back to "?" when metadata is missing a key', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'status_changed', metadata: {'from': 'new'}),
        myMembershipId: _me,
      );
      expect(entry.text, 'Status: Baru → ?');
    });

    test('lead_assigned to the viewer renders "Ditugaskan ke Anda"', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'lead_assigned', metadata: {'to': _me}),
        myMembershipId: _me,
      );
      expect(entry.text, 'Ditugaskan ke Anda');
    });

    test('lead_assigned to someone else never claims a resolved name', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'lead_assigned', metadata: {'to': _other}),
        myMembershipId: _me,
      );
      expect(entry.text, 'Ditugaskan ke Anggota tim lain');
    });

    test('lead_assigned with both from/to renders the move', () {
      final entry = activityToTimelineEntry(
        _activity(
          type: 'lead_assigned',
          metadata: {'from': _other, 'to': _me},
        ),
        myMembershipId: _me,
      );
      expect(entry.text, 'Dipindahkan dari Anggota tim lain ke Anda');
    });

    test('lead_unassigned', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'lead_unassigned', metadata: {'from': _me}),
        myMembershipId: _me,
      );
      expect(entry.text, 'Dilepas dari Anda');
    });

    test('lead_converted renders a static line', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'lead_converted'),
        myMembershipId: _me,
      );
      expect(entry.text, 'Dikonversi menjadi customer');
    });

    test('task_created includes the title when present', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'task_created', metadata: {'title': 'Follow up'}),
        myMembershipId: _me,
      );
      expect(entry.text, 'Task dibuat: Follow up');
    });

    test('task_created without a title falls back to a plain line', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'task_created'),
        myMembershipId: _me,
      );
      expect(entry.text, 'Task dibuat');
    });

    test('task_completed renders a static line', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'task_completed'),
        myMembershipId: _me,
      );
      expect(entry.text, 'Task diselesaikan');
    });

    for (final type in ['note_added', 'call_logged', 'whatsapp_opened']) {
      test('$type is human-authored, renders body verbatim, author "Anda" for self', () {
        final entry = activityToTimelineEntry(
          _activity(type: type, actorMembershipId: _me, body: 'halo'),
          myMembershipId: _me,
        );
        expect(entry.isHuman, isTrue);
        expect(entry.text, 'halo');
        expect(entry.authorName, 'Anda');
      });

      test('$type author is generic, never a resolved name, for another actor', () {
        final entry = activityToTimelineEntry(
          _activity(type: type, actorMembershipId: _other, body: 'halo'),
          myMembershipId: _me,
        );
        expect(entry.authorName, 'Anggota tim lain');
      });

      test('$type falls back to empty body, never null/"null" text', () {
        final entry = activityToTimelineEntry(
          _activity(type: type, actorMembershipId: _me),
          myMembershipId: _me,
        );
        expect(entry.text, '');
      });
    }

    test('an unrecognized type never throws — renders the raw type instead', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'something_new'),
        myMembershipId: _me,
      );
      expect(entry.isHuman, isFalse);
      expect(entry.text, 'something_new');
    });

    test('null actor renders "Seseorang", not "Anda"', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'note_added', body: 'x'),
        myMembershipId: _me,
      );
      expect(entry.authorName, 'Seseorang');
    });

    test('myMembershipId null never matches — everyone renders as generic, never "Anda"', () {
      final entry = activityToTimelineEntry(
        _activity(type: 'note_added', actorMembershipId: _me, body: 'x'),
        myMembershipId: null,
      );
      expect(entry.authorName, 'Anggota tim lain');
    });
  });

  group('lostReasonDisplayLabel', () {
    test('returns null for a null reason', () {
      expect(lostReasonDisplayLabel(null), isNull);
    });

    test('returns the Indonesian label for a known reason', () {
      expect(lostReasonDisplayLabel('price'), 'Harga');
    });
  });
}
